src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder.cs

```diff
@@ -38,17 +43,44 @@ internal static class DependentTypeFinder
         private static readonly ObjectPool<SymbolAndProjectIdSet> s_setPool = new ObjectPool<SymbolAndProjectIdSet>(
             () => new SymbolAndProjectIdSet(SymbolAndProjectIdComparer<INamedTypeSymbol>.SymbolEquivalenceInstance));
 
+        // Caches from a types to their related types (in the context of a specific solution).
+        // Kept as a cache so that clients who make many calls into us won't end up computing
+        // the same data over and over again.  Will be let go the moment the solution they're
+        // based off of is no longer alive.
+
+        private static readonly RelatedTypeCache s_typeToImmediatelyDerivedClassesMap = new RelatedTypeCache();
+        private static readonly RelatedTypeCache s_typeToTransitivelyDerivedClassesMap = new RelatedTypeCache();
+        private static readonly RelatedTypeCache s_typeToTransitivelyImplementingTypesMap = new RelatedTypeCache();
+        private static readonly RelatedTypeCache s_typeToImmediatelyDerivedAndImplementingTypesMap = new RelatedTypeCache();
+
+        public static async Task<ImmutableArray<SymbolAndProjectId<INamedTypeSymbol>>> FindTypesFromCacheOrComputeAsync(
+            INamedTypeSymbol type,
+            Solution solution,
+            RelatedTypeCache cache,
+            Func<CancellationToken, Task<ImmutableArray<SymbolAndProjectId<INamedTypeSymbol>>>> findAsync,
+            CancellationToken cancellationToken)
+        {
+            var dictionary = cache.GetOrCreateValue(solution);
+            var lazy = dictionary.GetOrAdd(type, new AsyncLazy<ImmutableArray<SymbolAndProjectId<INamedTypeSymbol>>>(
```

Comment:
AsyncLazy properly handles this.  An individual client can cancel if htey want.  AsyncLazy only cancels if all clients cancel.  If AsyncLazy cancels, it can still compute the value later if some new client comes in and requests it.